File: src/main/java/br/com/gabrielalmir/gotcha/service/UrlService.java
package br.com.gabrielalmir.gotcha.service;

import br.com.gabrielalmir.gotcha.dto.ShortenUrlRequest;
import br.com.gabrielalmir.gotcha.dto.ShortenUrlResponse;
import br.com.gabrielalmir.gotcha.entity.UrlEntity;
import br.com.gabrielalmir.gotcha.repository.UrlRepository;
import br.com.gabrielalmir.gotcha.utils.Base62Converter;
import br.com.gabrielalmir.gotcha.utils.Snowflake;
import jakarta.servlet.http.HttpServletRequest;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;

@Service
public class UrlService {
    final UrlRepository urlRepository;

    public UrlService(UrlRepository urlRepository) {
        this.urlRepository = urlRepository;
    }

    public ShortenUrlResponse shortenUrl(ShortenUrlRequest request, HttpServletRequest servletRequest) {
        var id = Base62Converter.toBase62(new Snowflake(1, 1).nextId());

        var entity = new UrlEntity(id, request.url(), LocalDateTime.now().plusMinutes(1));
        urlRepository.save(entity);

        var redirectUrl = servletRequest.getRequestURL().toString().replace("shorten-url", id);
        return new ShortenUrlResponse(redirectUrl);
    }

    public UrlEntity getUrl(String id) {
        return urlRepository.findById(id).orElse(null);
    }
}

File: src/services/url.service.ts
import { Request } from "express";
import { IUrlModel, UrlModel } from "../models/url.model";
import { Base62Converter } from "../utils/base62converter";
import { Snowflake } from "../utils/snowflake";

export class UrlService {
    private snowflake: Snowflake;

    constructor() {
        this.snowflake = new Snowflake(1, 1);
    }

    public async shortenUrl(requestUrl: string, request: Request) {
        const snowflakeId = this.snowflake.nextId();
        const id = Base62Converter.toBase62(snowflakeId);

        const entity = new UrlModel({
            _id: id,
            fullUrl: requestUrl,
            expiresAt: new Date(Date.now() + 1000 * 60 * 1),
        });

        await entity.save();

        const { protocol } = request;
        const hostname = request.get('host');
        const redirectUrl = `${protocol}://${hostname}/${id}`;

        return redirectUrl;
    }

    public async getUrl(id: string): Promise<IUrlModel | null> {
        return await UrlModel.findById(id);
    }
}
